benches/bin/utils.rs

```diff
@@ -0,0 +1,77 @@
+extern crate rand;
+
+use std::sync::atomic::{ATOMIC_U64_INIT, AtomicU64, Ordering};
+use std::time::Duration;
+use tikv::coprocessor::codec::table::{encode_index_seek_key, encode_row_key};
+use tikv::util::codec::number::NumberEncoder;
+
+use rand::Rng;
+
+#[inline]
+pub fn next_ts() -> u64 {
+    static CURRENT: AtomicU64 = ATOMIC_U64_INIT;
+    CURRENT.fetch_add(1, Ordering::SeqCst)
+}
+
+/// Generate `count` row keys that all with the specified `table_id`.
+///
+/// Row id will start with `start_id` and increment
+pub fn generate_row_keys(table_id: i64, start_id: i64, count: usize) -> Vec<Vec<u8>> {
+    let mut result = Vec::with_capacity(count);
+    for i in (start_id)..(start_id + count as i64) {
+        let mut handle = Vec::with_capacity(8);
+        handle.encode_i64(i as i64).unwrap();
+        let key = encode_row_key(table_id, &handle);
+        result.push(key);
+    }
+    result
+}
+
+
+/// Generate `count` unique index keys that all with the specified `table_id`, `index_id`,
+/// and `value_len`.
+pub fn generate_unique_index_keys(
+    table_id: i64,
+    index_id: i64,
+    value_len: usize,
+    count: usize,
+) -> Vec<Vec<u8>> {
+    let mut result = Vec::with_capacity(count);
+    for _ in 0..count {
+        result.push(encode_index_seek_key(
+            table_id,
+            index_id,
+            &vec![0u8; value_len],
+        ));
+    }
+    result
+}
+
+pub fn shuffle<T>(data: &mut [T]) {
+    let mut rng = rand::thread_rng();
+    for i in 0..(data.len()) {
+        let j = rng.gen_range(i, data.len());
+        data.swap(i, j);
+    }
+}
+
+/// Run `job` for `iterations` times, and return a Vec containing nanoseconds of each turn `job`
+/// returns.
+///
+/// `job` must return a `Duration` type. Please calculate time cost of what you want to bench
+/// manually ant return it.
+pub fn record_time<F>(mut job: F, iterations: u32) -> Vec<u64>
```

Comment:
what is this used for? why not return `Vec<Duration>`?

Reply:
Sometimes we want to bench some codes, and some initialization should be done before running these code. There may be a situation that, these code's side effect makes us not able run them several time but initialize only once. So here I let `job` record what it want and skip what it doesn't want. This function simply collect their result and return them.

Reply:
sorry, I still don't understand. If you want to support "initialize only once", why not using standard sync.Once?

Reply:
What I mean is this. Consider this common benchmark code:
```
// Initialization
b.iter(|| {
    // code to be benched
});
```
It does initialization once, and `b.iter` loops given function several times to record its time cost. If the function does some side effect that may affect time cost of the next iteration, which influences the precision of the result. We don't want to do this either:
```
b.iter(|| {
    // Initialization
    // code to be benched
});
```
which takes the unwanted time cost of initialization into calculation.
My idea is that, we can do this:
```
record_time(|| {
    // Initialization
    let t = Duration::now;
    // code to be benched
    t.elapsed()
}, iteration_times);
```
So that we can do initialization for every iteration, without calculating time cost of initialization code into bench results.

Reply:
so why not moving Initialization out of the iter?

Reply:
Sometimes the code we want to bench can't be simply repeated, unless we reset it before the next time it runs. So we want to do initialization for every iteration without calculating initialization time.

Reply:
Another question is that we can't specify number of iteration times to bencher in libtest. Sometimes it may be a trouble.

Reply:
Can you provide a case that needs this functionality?

Reply:
If you want to do this, it is better to provide a benchmark class which can first set up and warm up the data, then benchmark with specified times.